fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ScalarFunctionValue.java

```diff
@@ -0,0 +1,332 @@
+            if (resultType.isUnresolved()) {
+                resultType = argType;
+            } else if (!argType.isUnresolved()) {
+                resultType = Type.maximumType(resultType, argType);
+            }
+            SemanticException.check(resultType != null, SemanticException.ErrorCode.INCOMPATIBLE_TYPE);
+        }
+
+        final Optional<ScalarFunction> scalarFunction = Enums.getIfPresent(ScalarFunction.class, builtInFunction.getFunctionName().toUpperCase(Locale.getDefault())).toJavaUtil();
+        Verify.verify(scalarFunction.isPresent());
+
+        final PhysicalOperator physicalOperator = getOperatorMap().get(Pair.of(scalarFunction.get(), resultType.getTypeCode()));
+
+        Verify.verifyNotNull(physicalOperator, "unable to encapsulate scalar function due to type mismatch(es)");
+
+        List<Value> promotedArgs = new ArrayList<>();
+        for (final var arg: arguments) {
+            promotedArgs.add(PromoteValue.inject((Value) arg, resultType));
+        }
+        return new ScalarFunctionValue(physicalOperator, promotedArgs);
+    }
+
+    private static Map<Pair<ScalarFunction, TypeCode>, PhysicalOperator> computeOperatorMap() {
+        final ImmutableMap.Builder<Pair<ScalarFunction, TypeCode>, PhysicalOperator> mapBuilder = ImmutableMap.builder();
+        for (final PhysicalOperator operator : PhysicalOperator.values()) {
+            mapBuilder.put(Pair.of(operator.getScalarFunction(), operator.getResultType()), operator);
+        }
+        return mapBuilder.build();
+    }
+
+    /**
+     * The {@code greatest} function.
+     */
+    @AutoService(BuiltInFunction.class)
+    public static class GreatestFn extends BuiltInFunction<Value> {
+        public GreatestFn() {
+            super("greatest",
+                    ImmutableList.of(), new Type.Any(), ScalarFunctionValue::encapsulate);
+        }
+    }
+
+    /**
+     * The {@code least} function.
+     */
+    @AutoService(BuiltInFunction.class)
+    public static class LeastFn extends BuiltInFunction<Value> {
+        public LeastFn() {
+            super("least",
+                    ImmutableList.of(), new Type.Any(), ScalarFunctionValue::encapsulate);
+        }
+    }
+
+    /**
+     * The {@code coalesce} function.
+     */
+    @AutoService(BuiltInFunction.class)
+    public static class CoalesceFn extends BuiltInFunction<Value> {
+        public CoalesceFn() {
+            super("coalesce",
+                    ImmutableList.of(), new Type.Any(), ScalarFunctionValue::encapsulate);
+        }
+    }
+
+    /**
+     * Logical operator.
+     */
+    public enum ScalarFunction {
+        GREATEST,
+        LEAST,
+        COALESCE;
+    }
+
+    /**
+     * Physical operators.
+     */
+    @VisibleForTesting
+    public enum PhysicalOperator {
+        GREATEST_INT(ScalarFunction.GREATEST, TypeCode.INT, args -> greatest(args, Integer.class)),
```

Comment:
Maybe you should read this comment after the ones further down but the idea was to hard code the physical action here, e.g. this one would be doing hard-coded integer comparisons, not delegate to `compare()`